### PR TITLE
Add .gnu_debuglink parsing support to ElfFile

### PR DESCRIPTION
**Pull request by beckerhe**
This adds parsing support for the .gnu_debuglink section in ELF files.

Bug: http://b/178999632
Test: Ran unit tests

### CODE REVIEW

File: src/ElfUtils/ElfFile.cpp [hunk 1]
```diff
@@ -65,8 +67,47 @@ class ElfFileImpl : public ElfFile {
...
+template <typename ElfT>
+ErrorMessageOr<GnuDebugLinkInfo> ReadGnuDebuglinkSection(
+    const typename ElfT::Shdr& section_header, const llvm::object::ELFFile<ElfT>& elf_file) {
+  llvm::Expected<llvm::ArrayRef<uint8_t>> contents_or_error =
+      elf_file.getSectionContents(&section_header);
+  if (!contents_or_error) return ErrorMessage("Could not obtain contents.");
+
+  const llvm::ArrayRef<uint8_t>& contents = contents_or_error.get();
+
+  if (contents.size() < 4) return ErrorMessage{"Section is too short."};
```
Comment (dimitry-) [inline src/ElfUtils/ElfFile.cpp:82]: is this supposed to be sizeof(ChecksumType)?

Comment (dimitry-) [reply src/ElfUtils/ElfFile.cpp:82]: and probably <= because of sizeof(path) is minimum 1 ('\0')

File: src/ElfUtils/ElfFile.cpp [hunk 2]
```diff
@@ -65,8 +67,47 @@ class ElfFileImpl : public ElfFile {
...
+  }
+
+  if (path.size() > contents.size() - 4) return ErrorMessage{"No CRC32 checksum found"};
+
+  constexpr llvm::support::endianness kEndianness = ElfT::TargetEndianness;
+  using ChecksumType = uint32_t;
+  const auto checksum_storage = contents.slice(contents.size() - sizeof(ChecksumType));
+  constexpr size_t kAlignment = 1;
+  const auto reference_crc =
+      llvm::support::endian::read<ChecksumType, kAlignment>(checksum_storage.data(), kEndianness);
```
Comment (dimitry-) [inline src/ElfUtils/ElfFile.cpp:103]: while this is technically correct, I do not believe big-endian architecture will see the light of day in observable future.

personal preference: static_assert for LE somewhere and forget about endianness

**Review by dimitry- — APPROVED** [review]
otherwise lgtm

File: src/ElfUtils/ElfFile.cpp (hunk 1 above)
Comment (beckerhe) [reply src/ElfUtils/ElfFile.cpp]: Yes. Done. Thanks.

File: src/ElfUtils/ElfFile.cpp [hunk 3]
```diff
@@ -65,8 +67,52 @@ class ElfFileImpl : public ElfFile {
...
+
+  const llvm::ArrayRef<uint8_t>& contents = contents_or_error.get();
+
+  using ChecksumType = uint32_t;
+  constexpr size_t kMinimumPathLength = 1;
+  if (contents.size() < kMinimumPathLength + sizeof(ChecksumType)) {
+    return ErrorMessage{"Section is too short."};
+  }
+
+  if (contents.size() > 100 * 1024) {
```
Comment (florian-kuebler) [inline src/ElfUtils/ElfFile.cpp:88]: Can you also create a constant for this?

File: src/ElfUtils/ElfFile.cpp (hunk 2 above)
Comment (beckerhe) [reply src/ElfUtils/ElfFile.cpp]: Well there is MIPS, and technically ARM. :sweat_smile: 

But jokes aside, I see your point. Changed!

**Review by florian-kuebler — APPROVED** [review]
LGTM, thanks!

File: src/ElfUtils/ElfFileTest.cpp [hunk 4]
```diff
@@ -225,3 +225,26 @@ TEST(ElfFile, LineInfoNoDebugInfo) {
...
+TEST(ElfFile, HasGnuDebugLink) {
+  const std::filesystem::path file_path =
+      orbit_base::GetExecutableDir() / "testdata" / "hello_world_elf_with_gnu_debuglink";
+
+  auto hello_world = ElfFile::Create(file_path);
+  ASSERT_TRUE(hello_world) << hello_world.error().message();
+
+  EXPECT_TRUE(hello_world.value()->HasGnuDebuglink());
+  EXPECT_TRUE(hello_world.value()->GetGnuDebugLinkInfo().has_value());
+  EXPECT_EQ(hello_world.value()->GetGnuDebugLinkInfo()->path.string(), "hello_world_elf.debug");
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply src/ElfUtils/ElfFileTest.cpp:249]: We can't. The code is not written yet.

[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/ElfUtils/ElfFileTest.cpp:249]: Okay, makes sense :)

File: src/ElfUtils/include/ElfUtils/ElfFile.h [hunk 5]
```diff
@@ -21,6 +21,14 @@
 
 namespace orbit_elf_utils {
 
+struct GnuDebugLinkInfo {
+  std::filesystem::path path;
+
+  // That's a CRC32 checksum of the file contents of the separate debuginfo file.
+  // Check the ELF documentation for the exact polynomial and the initial value.
```
[the remark this answers is not in the document]
Comment (beckerhe) [reply src/ElfUtils/include/ElfUtils/ElfFile.h:28]: Done.

File: src/ElfUtils/ElfFile.cpp (hunk 3 above)
Comment (beckerhe) [reply src/ElfUtils/ElfFile.cpp]: Done.

**Pull request merged**